### PR TITLE
BTRS algorithm for fast/efficient binomial sampling

### PR DESCRIPTION
**Pull request by jamestwebber**
As a fix for #20343 I tried my hand at implementing the BTRS algorithm in `ATen/native` (basing my code on the TensorFlow implementation).

This was done almost entirely in cargo-cult style, trying to mimic the structure of similar methods in `aten/src/ATen/native/Distributions.cpp` etc. So there's probably stuff wrong with how I did things!

Changes are:
 - addition of `stirling_approx_tail`, `binomial_inversion`, and `btrs` functions to `aten/src/ATen/native/Distributions.h`
 - added `ceil` and `log1p` to `c10/cuda/CUDAMathCompat.h` because they are used in the above
 - added `_s_binomial_cpu` to `aten/src/ATen/native/Distributions.cpp`
 - added `_s_binomial_cuda` and `binomial_cuda_kernel` to `aten/src/ATen/native/cuda/Distributions.cu`
 - changed `torch/distributions/binomial.py` to use the new function instead of the huge-bernoulli strategy

I've tested the implementation on both CPU and CUDA and it works, although I have not tested that it generates the correct distribution. As expected it's much faster (~100x for CPU) than the old version, at least when `total_count` is high:

Previously, on my laptop:
```
counts = torch.randint(10, 1000, [1000,1000])
p = 0.5 * torch.ones(1000, 1000)

%timeit dist.binomial.Binomial(total_count=counts, probs=p).sample()
12.2 s ± 685 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Now:
```
%timeit dist.binomial.Binomial(total_count=counts, probs=p).sample()
90.1 ms ± 1.32 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

### CODE REVIEW

**Comment by jamestwebber** [comment]
[Reference for BTRS](https://doi.org/10.1080/00949659308811496). There is a faster algorithm in the paper but it is more complicated to implement.

The TensorFlow code was released under an Apache license and I left a comment to that effect in `Distributions.h`. I don't know if that is sufficient though.

**Comment by kostmo** [comment]
## CircleCI build failures summary
As of commit 69d9452:
* **1/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=69d94521ce6e26370c708c6adbdab42d0f43490d).

### 1 failure *not* recognized by patterns:
| Job | Step | Status |
| --- | --- | --- |
| CircleCI <sup>pytorch-linux-xenial-py3-clang5-android-ndk-r19c-gradle-build-x86_32</sup> | <sup>pytorch android gradle build only x86_32 (for PR)</sup> | New in PR |


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

**Comment by jamestwebber** [comment]
I'm not sure what do about the merge conflict. It's because of HIP-related changes to `Distributions.h` made in #31200 by @serhaty, which conflicts with my addition of a couple functions. I don't understand the HIPifying process so I don't know how to repeat the same exercise there–presumably it should be possible.

**Comment by jamestwebber** [comment]
I only noticed today that there are test failures. The first failure is because the variance of one sample is off by 0.0238 and the tolerance is 0.02&mdash;it's using the binomial inversion for that test and I'm not sure why it's off by that much. If someone else can verify that the code looks right that'd be great.

I understand the second error less, but it seems like something to do with how the new function is JITted.

**Comment by jamestwebber** [comment]
As far as I can tell the failure in `test_binomial_vectorized_count` is a fluke–if I run that test 100 times it fails once, and if I set `seed` to 0 it _happens_ to fail on the first iteration.

I don't know what the other failure means or how to fix it.

**Comment by fritzo** [comment]
> the failure ... is a fluke

Great, thanks for looking into this closely. Can you add prefix `test_sample()` with
```diff
  def test_sample(self):
      for Dist, keys, values, sample in self._examples():
+         if Dist is Binomial:
+             set_rng_seed(1)   # see Note [Randomized statistical tests]
```
Generally these statistical tests will all be a bit flakey, so we need to fix seeds, and occasionally fix seeds to something other than 0. With high probability, we should be able to set all seeds to either 0 or 1, which is the convention we've followed in test_distributions.py. If you need to increment the seed multiple times that suggests your test is truely fails.

**Comment by jamestwebber** [comment]
I think you meant `test_binomial_vectorized_count`? That function starts with `set_rng_seed(0)`, when I change that to 1 it passes.

The other test still failed with this error: 
```
Traceback (most recent call last):
  File "test/test_distributions.py", line 4412, in test_sample
    self.assertTrue(any(n.isNondeterministic() for n in traced_f.graph.nodes()))
AssertionError: False is not true
```
Which I infer means that nothing in the Binomial example is non-deterministic. Now that I actually bother to search the codebase for that function, I'm guessing that `aten::binomial` should be added to [`torch/csrc/jit/ir.cpp`](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/ir.cpp#L916) in `Node::isNondeterministic`?

**Comment by neerajprad** [comment]
@jamestwebber - Thanks for this PR, and sorry for being late on this. I would really like to help get this merged asap. I'll take a closer look at the changes and the math. Just to understand the current status - the existing test failures seem unrelated, so once the latest master is merged in, my understanding is that this should be good as far as existing tests go. Is that correct?

**Comment by jamestwebber** [comment]
I think so but I'm not sure. The only relevant issue is the merge conflict
related to HIP, because I needed some additional math functions and I'm not
sure how that works with HIP.

**Comment by ezyang** [comment]
@neerajprad I'm very happy to defer to you for full review of the PR. Let me see if I can unblock the HIP stuff.

**Comment by ezyang** [comment]
I think I fixed the HIP changes

**Comment by neerajprad** [comment]
Thanks for your help, @ezyang. I'm reviewing the rest of the PR now.

**Review by neerajprad — COMMENTED** [review]
This looks great, I have checked against the reference tf implementation. My only comment is about adding a test for the sampler which we currently lack.

File: aten/src/ATen/native/Distributions.h
Comment (t-vi) [reply aten/src/ATen/native/Distributions.h]: I'm not even sure that we use Apache-licensed code in PyTorch, in fact, I think we cannot. That said, thankfully I'm no authority here.

Comment (fritzo) [reply aten/src/ATen/native/Distributions.h]: Caffe2 is [licensed](https://github.com/facebookarchive/caffe2/blob/master/LICENSE) under Apache 2, so it should be fine to include other Apache 2 code

File: test/test_distributions.py
Comment (jamestwebber) [reply test/test_distributions.py]: I'm away from my work computer until next week but will be happy to do it then.

**Comment by jamestwebber** [comment]
I've spent the day trying to convince pytorch to compile on at least one platform, to no avail. I just added the test directly on GH so I wasn't able to run it locally, hopefully CI doesn't complain.

**Comment by jamestwebber** [comment]
It looks like the merge conflict got a little mangled...there were `cuda` defines in the `hip` section and not every function was present. I fixed it but I don't know if the `HIP` stuff works

**Review by neerajprad — APPROVED** [review]
LGTM. @ezyang - I defer to you for the changes in `Distributions.h` and using the Apache 2 licensed code, should it be an issue. cc. @soumith.

**Comment by soumith** [comment]
wrt license, we're good. LICENSE already specifies both BSD3 and Apache 2, based on the Caffe2 code that we pulled in -- parts of the codebase are BSD and other parts are Apache2. So adding in more apache2 with correct inline copyright headers seems good.

**Comment by neerajprad** [comment]
The changes in `distributions.h` seem in accordance with the rest of the file, and the test failures are unrelated, so I'm merging this.  cc @ezyang. 

@pytorchbot - merge this please.

**Comment by neerajprad** [comment]
This probably doesn't work anymore, or I don't have authorization to merge. @ezyang, @zou3519 - could you help?

**Comment by zou3519** [comment]
@ailzhang FYI the XLA build failed on this but I am not sure if it is related or not. I'm going to merge this PR (let me know if I shouldn't)

**Comment by zou3519** [comment]
Hmm, this might be causing some correctness issues on some internal tests. I'll try to find a minimal test case that shows the issues and report back

**Comment by mruberry** [comment]
The PyTorch/XLA failure is not related to this PR. Sorry for the noise (and thank you for asking about it!)

**Comment by mbabadi** [comment]
@neerajprad Hi Neeraj, is there progress in merging this PR? I am anxiously waiting to try it. Your NumPyro implementation is great btw! :-)

**Comment by neerajprad** [comment]
@zou3519, @mruberry - I thought that this was already merged. Were there any issues internally and this PR was reverted?

**Review by apaszke — COMMENTED** [review]
This looks good to me, except for the licensing issues. I think we should really better understand if Apache-licensed code can be relicensed as BSD3 (I don't think so, and [others think so too](https://softwareengineering.stackexchange.com/a/40571)!).

**Comment by mbabadi** [comment]
I just wanted to add a quick side note that I compiled this branch, thoroughly tested it, and all I saw was ... 50x speedup! thank you @jamestwebber for this great addition and everyone involved in reviewing this PR. I hope the licensing issues can be resolved at some point.

**Comment by ezyang** [comment]
Still some Windows portability problems

```
C:/Users/circleci/project/aten/src/ATen/native/cuda/Distributions.cu(75): error: namespace "at::cuda" has no member "CUDA_tensor_apply3"
          detected during instantiation of "void <unnamed>::binomial_cuda_kernel<scalar_t>(at::Tensor &, const at::Tensor &, const at::Tensor &, std::pair<uint64_t, uint64_t>) [with scalar_t=double]" 
(195): here

C:/Users/circleci/project/aten/src/ATen/native/cuda/Distributions.cu(75): error: type name is not allowed
          detected during instantiation of "void <unnamed>::binomial_cuda_kernel<scalar_t>(at::Tensor &, const at::Tensor &, const at::Tensor &, std::pair<uint64_t, uint64_t>) [with scalar_t=double]" 
(195): here

C:/Users/circleci/project/aten/src/ATen/native/cuda/Distributions.cu(75): error: type name is not allowed
          detected during instantiation of "void <unnamed>::binomial_cuda_kernel<scalar_t>(at::Tensor &, const at::Tensor &, const at::Tensor &, std::pair<uint64_t, uint64_t>) [with scalar_t=double]" 
(195): here

C:/Users/circleci/project/aten/src/ATen/native/cuda/Distributions.cu(75): error: type name is not allowed
          detected during instantiation of "void <unnamed>::binomial_cuda_kernel<scalar_t>(at::Tensor &, const at::Tensor &, const at::Tensor &, std::pair<uint64_t, uint64_t>) [with scalar_t=double]" 
(195): here

C:/Users/circleci/project/aten/src/ATen/native/cuda/Distributions.cu(75): error: namespace "at::cuda" has no member "CUDA_tensor_apply3"
          detected during instantiation of "void <unnamed>::binomial_cuda_kernel<scalar_t>(at::Tensor &, const at::Tensor &, const at::Tensor &, std::pair<uint64_t, uint64_t>) [with scalar_t=float]" 
(195): here

C:/Users/circleci/project/aten/src/ATen/native/cuda/Distributions.cu(75): error: type name is not allowed
          detected during instantiation of "void <unnamed>::binomial_cuda_kernel<scalar_t>(at::Tensor &, const at::Tensor &, const at::Tensor &, std::pair<uint64_t, uint64_t>) [with scalar_t=float]" 
(195): here

C:/Users/circleci/project/aten/src/ATen/native/cuda/Distributions.cu(75): error: type name is not allowed
          detected during instantiation of "void <unnamed>::binomial_cuda_kernel<scalar_t>(at::Tensor &, const at::Tensor &, const at::Tensor &, std::pair<uint64_t, uint64_t>) [with scalar_t=float]" 
(195): here

C:/Users/circleci/project/aten/src/ATen/native/cuda/Distributions.cu(75): error: type name is not allowed
          detected during instantiation of "void <unnamed>::binomial_cuda_kernel<scalar_t>(at::Tensor &, const at::Tensor &, const at::Tensor &, std::pair<uint64_t, uint64_t>) [with scalar_t=float]" 
(195): here

C:/Users/circleci/project/aten/src/ATen/native/cuda/Distributions.cu(75): error: namespace "at::cuda" has no member "CUDA_tensor_apply3"
          detected during instantiation of "void <unnamed>::binomial_cuda_kernel<scalar_t>(at::Tensor &, const at::Tensor &, const at::Tensor &, std::pair<uint64_t, uint64_t>) [with scalar_t=c10::Half]" 
(195): here

C:/Users/circleci/project/aten/src/ATen/native/cuda/Distributions.cu(75): error: type name is not allowed
          detected during instantiation of "void <unnamed>::binomial_cuda_kernel<scalar_t>(at::Tensor &, const at::Tensor &, const at::Tensor &, std::pair<uint64_t, uint64_t>) [with scalar_t=c10::Half]" 
(195): here

C:/Users/circleci/project/aten/src/ATen/native/cuda/Distributions.cu(75): error: type name is not allowed
          detected during instantiation of "void <unnamed>::binomial_cuda_kernel<scalar_t>(at::Tensor &, const at::Tensor &, const at::Tensor &, std::pair<uint64_t, uint64_t>) [with scalar_t=c10::Half]" 
(195): here

C:/Users/circleci/project/aten/src/ATen/native/cuda/Distributions.cu(75): error: type name is not allowed
          detected during instantiation of "void <unnamed>::binomial_cuda_kernel<scalar_t>(at::Tensor &, const at::Tensor &, const at::Tensor &, std::pair<uint64_t, uint64_t>) [with scalar_t=c10::Half]" 
(195): here
```

**Comment by ezyang** [comment]
Seems to be affecting HIP too.

**Comment by neerajprad** [comment]
@ezyang - is it be related to https://github.com/pytorch/pytorch/pull/34026?

**Comment by ezyang** [comment]
Yep.

**Comment by ezyang** [comment]
Think I fixed it

**Comment by ezyang** [comment]
Hi folks, some bad news: this is causing one of our internal tests to fail. In this test, it seems the mean of the distribution under a specific seed doesn't expect the theoretical mean. This tests comes from a private project on GitHub but it looks like we haven't open sourced it yet.

```
======================================================================
FAIL: test_beta_binomial_conjugate_run (tests.conjugate_tests.newtonian_monte_carlo_conjugate_test.SingleSiteNewtonianMonteCarloConjugateTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "[redacted]/tests/conjugate_tests/newtonian_monte_carlo_conjugate_test.py", line 17, in test_beta_binomial_conjugate_run
    self.beta_binomial_conjugate_run(nw, num_samples=150, delta=0.15)
  File "[redacted]/tests/conjugate_tests/abstract_conjugate.py", line 208, in beta_binomial_conjugate_run
    self._compare_run(moments, mh, num_chains, num_samples, delta, random_seed)
  File "[redacted]/conjugate_tests/abstract_conjugate.py", line 187, in _compare_run
    abs((mean - expected_mean).sum().item()), 0, delta=delta
AssertionError: 0.26111966371536255 != 0 within 0.15 delta (0.26111966371536255 difference)
```

Excerpts of the code:

```
    def test_beta_binomial_conjugate_run(self):
        nw = SingleSiteNewtonianMonteCarlo()
        self.beta_binomial_conjugate_run(nw, num_samples=150, delta=0.15)  # failing line
        nw = SingleSiteNewtonianMonteCarlo(True, 5.0, 1.0)
        self.beta_binomial_conjugate_run(nw, num_samples=150, delta=0.15)
```

```
    def beta_binomial_conjugate_run(
        self,
        mh: AbstractInference,
        num_chains: int = 1,
        num_samples: int = 1000,
        delta: float = 0.05,
        random_seed: Optional[int] = 17,
    ):
        """
        Tests the inference run for a small beta binomial model.

        :param mh: inference algorithm
        :param num_samples: number of samples
        :param num_chains: number of chains
        :param delta: delta to check against expected results
        :param random_seed: seed for pytorch random number generator
        """
        moments = self.compute_beta_binomial_moments()
        self._compare_run(moments, mh, num_chains, num_samples, delta, random_seed)

    def _compare_run(
        self,
        moments: Tuple[Tensor, Tensor, List[RVIdentifier], Dict[RVIdentifier, Tensor]],
        mh: AbstractInference,
        num_chains: int,
        num_samples: int,
        delta: float,
        random_seed: Optional[int],
    ):  
        expected_mean, expected_std, queries, observations = moments
        
        if random_seed is not None:
            torch.manual_seed(random_seed)
        
        predictions = mh.infer(queries, observations, num_samples, num_chains)
        for i in range(predictions.get_num_chains()):
            mean, _ = self.compute_statistics(predictions.get_chain(i)[queries[0]])
            # pyre-fixme[16]: `AbstractConjugateTests` has no attribute
            #  `assertAlmostEqual`.
            self.assertAlmostEqual(
                abs((mean - expected_mean).sum().item()), 0, delta=delta
            )

    def compute_statistics(self, predictions: Tensor) -> Tuple[Tensor, Tensor]:
        """ 
        Computes mean and standard deviation of a given tensor of samples.
        
        :param predictions: tensor of samples
        :returns: mean and standard deviation of the tensor of samples.
        """ 
        return (torch.mean(predictions, 0), torch.std(predictions, 0))

```

```
    def compute_beta_binomial_moments(
        self
    ) -> Tuple[Tensor, Tensor, List[RVIdentifier], Dict[RVIdentifier, Tensor]]:
        """
        Computes mean and standard deviation of a small beta binomial model.

        :return: expected mean, expected standard deviation, conjugate model
        queries and observations
        """
        alpha = tensor([2.0, 2.0])
        beta = tensor([1.0, 1.0])
        trials = tensor([1.0, 1.0])
        obs = tensor([1.0, 0.0])
        model = BetaBinomialModel(alpha, beta, trials)
        queries = [model.beta()]
        observations = {model.binomial(): obs}
        alpha = alpha + obs
        beta = beta - obs + trials
        expected_mean = alpha / (alpha + beta)
        # pyre-fixme[16]: `float` has no attribute `pow`.
        expected_std = (
            (alpha * beta) / (alpha + beta).pow(2.0) * (alpha + beta + 1.0)
        ).pow(0.5)
                
        return (expected_mean, expected_std, queries, observations)
```

```
import torch.distributions as dist
from torch import Tensor


class BetaBinomialModel(object):
    def __init__(self, alpha: Tensor, beta: Tensor, trials: Tensor): 
        self.alpha_ = alpha
        self.beta_ = beta
        self.trials_ = trials

    @sample
    def beta(self):
        return dist.Beta(self.alpha_, self.beta_)

    @sample
    def binomial(self):
        return dist.Binomial(self.trials_, self.beta())
```

I noted in the PR description that you haven't checked that the correct distribution is actually produced. Maybe it would be a good time to check this? This might just be noise, in which case I can just increase the tolerance, but I'd like someone who actually knows about this to check.

**Comment by neerajprad** [comment]
@ezyang - Thanks for sharing the test details, very interesting. Could you check if the test passes with more number of samples? This PR changes the binomial sampler. The model in the test constrains the binomial site to some observed value and tests the posterior probability over the beta site, so I'm unsure why the change in the sampler would affect the inference results since we won't be sampling from the Binomial distribution for this model, IIUC. 

@nimar - any ideas why this test might be failing? 🙂

**Comment by jamestwebber** [comment]
I did have to change the seed for one other test, this might just be another case of that. Does it fail every time?

**Comment by ezyang** [comment]
OK, I will try increasing the number of samples to start.

**Comment by ezyang** [comment]
I pumped up the samples 10x and the delta got worse:

```
======================================================================
FAIL: test_beta_binomial_conjugate_run (tests.conjugate_tests.newtonian_monte_carlo_conjugate_test.SingleSiteNewtonianMonteCarloConjugateTest)
----------------------------------------------------------------------
[redacted]
    abs((mean - expected_mean).sum().item()), 0, delta=delta
AssertionError: 1.032049298286438 != 0 within 0.15 delta (1.032049298286438 difference)
```

**Comment by ezyang** [comment]
@jamestwebber if I produce an OSS repro of this test failure would that be helpful?

**Comment by jamestwebber** [comment]
I won't be able to look at it for a bit (I'm writing this from an
airport...). Last time I tried I wasn't able to build pytorch so I might
not be able to test anything anyway. But it would probably help someone at
least.

**Comment by neerajprad** [comment]
@ezyang - I can take a look if you could share the failing test case.

File: aten/src/ATen/native/cuda/Distributions.cu
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/Distributions.cu:74]: I don't think this is correct, because each thread can call the lambda multiple times. If implemented this way, it will generate the same random number multiple times. 

I don't think there is an easy way to do distribution kernels with `at::native::gpu_kernel` due to the reason I mentioned above. And my solution is to write my own elementwise kernels. See: https://github.com/pytorch/pytorch/pull/34004/files

**Comment by ezyang** [comment]
Sorry about the delay, haven't had a chance to excise out the test case yet

File: aten/src/ATen/native/cuda/Distributions.cu
Comment (ezyang) [reply aten/src/ATen/native/cuda/Distributions.cu:74]: Thanks for the note, @zasdfgbnm. Unfortunately this won't explain the internal test failure (which is running CPU only)

**Comment by ezyang** [comment]
The test case is too big to extract. I'm going to reach out to the relevant internal team and see if they have any guidance.

**Comment by fritzo** [comment]
Is this ready for rebase? I heard the internal issue was resolved.

**Comment by ezyang** [comment]
Yes it was resolved. Rebase and I'll attempt land again! Though, I suppose we should also fix the issue @zasdfgbnm pointed out too.

**Comment by jamestwebber** [comment]
@fritzo are you going to take care of that? I'm not sure I understand/know how to fix the issue @zasdfgbnm raised.

**Comment by fritzo** [comment]
@jamestwebber sorry I don't understand the issue raised by @zasdfgbnm .

**Comment by zasdfgbnm** [comment]
@fritzo When calling the lambda, we need to make sure the `sequence` passed to `curand_init` be different for different lambda calls. But this is not true because the sequence is set as `blockIdx.x * blockDim.x + threadIdx.x`, and each thread can call the lambda multiple times.

**Comment by crcrpar** [comment]
@jamestwebber @fritzo @zasdfgbnm I'd like to work on this.

**Comment by jamestwebber** [comment]
> @jamestwebber @fritzo @zasdfgbnm I'd like to work on this.

Great to hear! I'm certainly not going to get in your way, way too busy with other work. 😄

**Comment by crcrpar** [comment]
@jamestwebber I really appreciate your initial great work.

As my takeover got merged, how about closing this?

**Comment by jamestwebber** [comment]
Sounds good to me, didn't know if I need to do that.

**Pull request closed without merging**
